### PR TITLE
clarify atomicity of deploy start

### LINKED ISSUE #8325: Deploy.execute does not report resource state when deploy result is None
It has an `if deploy_result is not None` condition. ~~However, this violates the contract of the `TaskManager`. The method must always be called. If need be the method's contract must be updated so that the notification becomes a NOOP, but I actually think it would be more appropriate to simply report a failure.~~ EDIT: the mandatory part is already outside of the if body. The relevant part is described below.

A more delicate instance of this problem exists [here](https://github.com/inmanta/inmanta-core/blame/652e14fc50e0f507cd51d15bfbf933e4beabfed4/src/inmanta/deploy/tasks.py#L148). The deploy start notification resulted in an error. Do we need to report that it finished? One observed case had the server register the resource as "deploying", with the result that it was stuck in this state.

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

Ensure we end the deploy if we start it, all part where in place, this is mostly docs 

closes #8325 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: src/inmanta/deploy/tasks.py [hunk 1]
```diff
@@ -120,41 +120,47 @@ async def execute(self, task_manager: "scheduler.TaskManager", agent: str, reaso
...
+            # From this point on, we HAVE to call report_resource_state to make the scheduler propagate state
 
             # The main difficulty off this code is exception handling
             # We collect state here to report back in the finally block
 
             # Full status of the deploy,
             # may be unset if we fail before signaling start to the server, will be set if we signaled start
             deploy_result: DeployResult | None = None
-            scheduler_deployment_result: state.DeploymentResult
+            scheduler_deployment_result: state.DeploymentResult = state.DeploymentResult.FAILED
```
Comment (jptrindade) [inline src/inmanta/deploy/tasks.py:131]: I remember @sanderr not wanting to initialize this value and to explicitly set it if the deploy were to fail. Like in line 154

Comment (wouterdb) [reply src/inmanta/deploy/tasks.py:131]: The question is: where do you catch that failure?

We caught all the ones we expect, but if something unexpected happens, do we prefer to crash the scheduler or report failure and continue

Comment (sanderr) [reply src/inmanta/deploy/tasks.py:131]: I think João may be referring to my preference for style a over style b in the snippet below
```python
# style a
x: int
try:
  do_stuff()
  x = 0
except Exception:
  x = 1

# style b
y: int = 1
try:
  do_stuff()
  y = 0
except Exception:
  pass
```

This is mostly a stylistic preference (semi-functional over updating the value), with a small part of confidence that we can't forget to set it in any of the branches.

But it remains mostly stylistic, and thus subjective, so I won't raise a fuss.

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/deploy/persistence.py [hunk 2]
```diff
@@ -44,6 +44,12 @@ class StateUpdateManager(abc.ABC):
 
     @abc.abstractmethod
     async def send_in_progress(self, action_id: UUID, resource_id: ResourceVersionIdStr) -> None:
+        """
+        This method sets the state to in_progress.
+
+        It is important that this method is atomic: if it fails, we assume the state to be not set and will not re-set it
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/deploy/persistence.py:50]: (the to_server version definitely was not atomic, this one is)

[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/deploy/persistence.py:50]: Right, it's not to_server anymore

**Review by sanderr — APPROVED** [review]

**Pull request closed without merging**